tests/testthat/test-string_boundary_linter.R

```diff
@@ -0,0 +1,155 @@
+  #   substring(s <- "abcdefg", 2L) == "efg" is not TRUE, but endsWith(s, "efg")
+  #   is. And if `s` contains strings of varying lengths, there's no equivalent.
+  expect_lint("substring(x, 2L)", NULL, string_boundary_linter())
+})
+
+test_that("string_boundary_linter blocks simple disallowed grepl() usages", {
+  expect_lint(
+    "grepl('^a', x)",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+  expect_lint(
+    "grepl('a$', x)",
+    rex::rex("Use endsWith() to detect a fixed terminal substring."),
+    string_boundary_linter()
+  )
+  # also get negation for free
+  expect_lint(
+    "!grepl('a$', x)",
+    rex::rex("Use endsWith() to detect a fixed terminal substring."),
+    string_boundary_linter()
+  )
+
+  # perl = TRUE doesn't matter
+  expect_lint(
+    "grepl('^a', x, perl = TRUE)",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+  # explicit FALSE (i.e., an explicit default) is ignored
+  expect_lint(
+    "grepl('^a', x, fixed = FALSE)",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+  expect_lint(
+    "grepl('^a', x, fixed = F)",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+})
+
+test_that("string_boundary_linter blocks simple disallowed str_detect() usages", {
+  expect_lint(
+    "str_detect(x, '^a')",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+  expect_lint(
+    "str_detect(x, 'a$')",
+    rex::rex("Use endsWith() to detect a fixed terminal substring."),
+    string_boundary_linter()
+  )
+})
+
+test_that("string_boundary_linter blocks disallowed substr()/substring() usage", {
+  expect_lint(
+    "substr(x, 1L, 2L) == 'ab'",
+    rex::rex("Use startsWith() to detect an initial substring."),
+    string_boundary_linter()
+  )
+  # end doesn't matter, just anchoring to 1L
+  expect_lint(
+    "substr(x, 1L, end) == 'ab'",
+    rex::rex("Use startsWith() to detect an initial substring."),
+    string_boundary_linter()
+  )
+  expect_lint(
+    "substring(x, nchar(x) - 4L, nchar(x)) == 'abcde'",
+    rex::rex("Use endsWith() to detect a terminal substring."),
+    string_boundary_linter()
+  )
+  # start doesn't matter, just anchoring to nchar(x)
+  expect_lint(
+    "substring(x, start, nchar(x)) == 'abcde'",
+    rex::rex("Use endsWith() to detect a terminal substring."),
+    string_boundary_linter()
+  )
+  # more complicated expressions
+  expect_lint(
+    "substring(colnames(x), start, nchar(colnames(x))) == 'abc'",
+    rex::rex("Use endsWith() to detect a terminal substring."),
+    string_boundary_linter()
+  )
+})
+
+test_that("plain ^ or $ are skipped", {
+  expect_lint('grepl("^", x)', NULL, string_boundary_linter())
+  expect_lint('grepl("$", x)', NULL, string_boundary_linter())
+})
```

Comment:
Can't we lint these and suggest `!is.na(x)`?

Reply:
Filed #1331 -- I didn't actually see this usage anywhere (it's very peculiar after all), the test was just covering bases when designing the linter. So I don't think it's important to include initially.